src/lib/comms/sol-http-common.c

```diff
@@ -74,52 +130,645 @@ sol_http_param_free(struct sol_http_param *params)
+        return -EINVAL;
+    }
+
+    //Not an URL
+    if (strncmp(auth.data, "//", 2))
+        return 0;
+
+    *port = -1;
+    auth.data += 2;
+    auth.len -= 2;
+
+    sep = memchr(auth.data, '@', auth.len);
+
+    if (sep) {
+        struct sol_str_slice info;
+        char *pass_sep;
+        info.data = auth.data;
+        info.len = sep - auth.data;
+        *user = info;
+        pass_sep = memchr(info.data, ':', info.len);
+        offset++;
+        if (pass_sep) {
+            user->len = pass_sep - user->data;
+            pass->data = pass_sep + 1;
+            pass->len = sep - pass->data;
+            offset++;
+        }
+        SOL_DBG("User: %.*s Pass: %.*s", SOL_STR_SLICE_PRINT(*user),
+            SOL_STR_SLICE_PRINT(*pass));
+        auth.data = sep + 1;
+        auth.len -= pass->len + user->len + 2;
+    }
+
+    if (!auth.len) {
+        SOL_WRN("Empty authority at: %.*s", SOL_STR_SLICE_PRINT(full_uri));
+        return -EINVAL;
+    }
+
+    host->data = auth.data;
+    //Ipv6
+    if (auth.data[0] == '[') {
+        sep = memchr(auth.data, ']', auth.len);
+        if (!sep) {
+            SOL_WRN("Could not find the ] character in the URI: %.*s",
+                SOL_STR_SLICE_PRINT(full_uri));
+            return -EINVAL;
+        }
+        host->len = sep - auth.data + 1;
+        auth.data = sep + 1;
+        auth.len -= host->len;
+    }
+
+    port_sep = memchr(auth.data, ':', auth.len);
+
+    fragment_sep = memchr(auth.data, '#', auth.len);
+    query_sep = memchr(auth.data, '?', auth.len);
+    path_sep = memchr(auth.data, '/', auth.len);
+
+    sep = _get_first_sep(query_sep, fragment_sep);
+    sep = _get_first_sep(sep, path_sep);
+
+    //Has port
+    if (port_sep) {
+        char *endptr;
+        if (!host->len)
+            host->len = port_sep - auth.data;
+        port_slice.data = port_sep + 1;
+        if (sep)
+            port_slice.len = sep - port_slice.data + 1;
+        else
+            port_slice.len = partial_uri.len - host->len -
+                user->len - pass->len - 2;
+        *port = sol_util_strtol(port_slice.data, &endptr, port_slice.len, 10);
+
+        if (endptr == port_slice.data) {
+            SOL_WRN("Could not convert the host port to integer. Port: %.*s",
+                SOL_STR_SLICE_PRINT(port_slice));
+            return -EINVAL;
+        }
+    } else if (sep)
+        host->len = sep - host->data;
+    else if (!sep)
+        host->len = auth.len;
+
+    _update_partial_uri(next,
+        port_slice.len + host->len + pass->len + user->len + offset);
+    SOL_DBG("Host: %.*s Port: %d", SOL_STR_SLICE_PRINT(*host), *port);
+    return 0;
+}
+
+static int
+_get_path(const struct sol_str_slice partial_uri,
+    const struct sol_str_slice full_uri,
+    struct sol_str_slice *next,
+    struct sol_str_slice *path)
+{
+    char *sep_query, *sep_fragment;
+
+    if (!partial_uri.len)
+        return 0;
+
+    if (partial_uri.data[0] == '#' || partial_uri.data[0] == '?')
```

Comment:
in these cases we have an empty path, should we make sure path points to an empty slice?